joomla/joomla-cms
# [4.0] No TLS transfer encryption for database connections if localhost

Pull Request for Issue # .

Follow-up to PR #26375 .

### Summary of Changes

With PR #26375 the possibility to use encrypted database connections was added to the CMS, and with PR #26888 it would be added to the installation process, too, if that is accepted.

Again thanks to @andrepereiradasilva for the implementation of these new features.

Because database connections default to socket connections in case if the database server is 'localhost' (which is of course more performant that a TCP/IP connection like e.g. to '127.0.0.1' or '::1'), and encryption is not possible with socket connections, this PR here changes the showon condition for the related fields in the server section of the Global Configuration form so that they are hidden when database server is 'localhost', and it adds an onchange JS to clear the fields when changing database server from something else to 'localhost'.

The JS added by this PR has the advantage that it can be used in future for other onchange events, too, e.g. if we want to clear the FTP related fields when disabling the FTP layer.

### Testing Instructions

1. On an installation of current 4.0-dev or last nightly build, go to Global Configuration, tab "Server", section "Database".
2. If field "Host" has a value different to "localhost", change it to "localhost".
3. Select option "Two-way encryption" of field "Connection Encryption".
4. Enter some values for the text fields related to database connection encryption as shown below in section "Actual result" in the screen shot.
Result: See section "Actual result" below.
5. Click the "Close" button in order to leave the form without saving the changes.
6. Apply the patch of this PR e.g. with [Patch Tester 3.0.0 Beta 4](https://github.com/joomla-extensions/patchtester/releases/tag/3.0.0-beta4) or [Patch Tester 4.0.0 Beta 2](https://github.com/joomla-extensions/patchtester/releases/tag/4.0.0-beta2) (if the latter works for you).
7. If not having used [Patch Tester 4.0.0 Beta 2](https://github.com/joomla-extensions/patchtester/releases/tag/4.0.0-beta2) with the CI option enabled:
Run `npm run build:js`.
8. Go again to Global Configuration, tab "Server", section "Database".
9. If field "Host" has a value different to "localhost", change it to "localhost".
Result: See screenshot 1 in section "Expected result" below.
10. Change the host to something different to "localhost".
Result: See screenshot 2 in section "Expected result" below.
11. Select option "Two-way encryption" of field "Connection Encryption".
12. Enter some values for the text fields related to database connection encryption.
Result: See screenshot 3 in section "Expected result" below.
13. Change the host back to "localhost".
Result: The options for database connection encryption are hidden again.
14. Change again the host to something different to "localhost".
Result: See screenshot 4 in section "Expected result" below.
15. Click the "Close" button in order to leave the form without saving the changes.

### Expected result

1. Database connection encryption related options are hidden if database host is "localhost".
![j4-db-encryption-localhost-after-patch_2](https://user-images.githubusercontent.com/7413183/69009693-3e95d400-0958-11ea-874c-719623877379.png)
2. The "Connection Encryption" field is shown if database host is different to "localhost".
![j4-db-encryption-localhost-after-patch_3](https://user-images.githubusercontent.com/7413183/69009749-c845a180-0958-11ea-8484-9d258c93b580.png)
3. If "Two-way encryption" is selected in field "Connection Encryption" you can enter values in the text fields shown in this case.
![j4-db-encryption-localhost-after-patch_1](https://user-images.githubusercontent.com/7413183/69009775-32f6dd00-0959-11ea-9d36-beecbf66d4c4.png)
4. If the hostname has been changed to "localhost", the options for database connection encryption are reset.
![j4-db-encryption-localhost-after-patch_4](https://user-images.githubusercontent.com/7413183/69009825-ab5d9e00-0959-11ea-8996-b2f7c84ded0c.png)

### Actual result

When the database host (field "Host") is "localhost", fields related to database connection encryption are shown depending on the value selected in field "Connection Envrytion". If "Two-way encryption" is selected, you can enter values in the text fields shown in this case.

![j4-db-encryption-localhost-before-patch](https://user-images.githubusercontent.com/7413183/69009657-dc3cd380-0957-11ea-8c8e-f52de0430e36.png)

### Documentation Changes Required

See PR #26375 .
